### PR TITLE
Use `TypeGuard` for `has` in Python 3.10 and above

### PR DESCRIPTION
**Pull request by layday**
# Summary

Conditionally defines `has` using `TypeGuard` in Python 3.10+ in the stub file.

xref: https://github.com/python-attrs/attrs/issues/987

# Pull Request Check List

- [x] Added **tests** for changed code.
  Our CI fails if coverage is not 100%.
- [ ] New features have been added to our [Hypothesis testing strategy](https://github.com/python-attrs/attrs/blob/main/tests/strategies.py).
- [ ] Changes or additions to public APIs are reflected in our type stubs (files ending in ``.pyi``).
    - [ ] ...and used in the stub test file `tests/typing_example.py`.
    - [ ] If they've been added to `attr/__init__.pyi`, they've *also* been re-imported in `attrs/__init__.pyi`.
- [ ] Updated **documentation** for changed code.
    - [ ] New functions/classes have to be added to `docs/api.rst` by hand.
    - [ ] Changes to the signature of `@attr.s()` have to be added by hand too.
    - [ ] Changed/added classes/methods/functions have appropriate `versionadded`, `versionchanged`, or `deprecated` [directives](http://www.sphinx-doc.org/en/stable/markup/para.html#directive-versionadded).
          Find the appropriate next version in our [``__init__.py``](https://github.com/python-attrs/attrs/blob/main/src/attr/__init__.py) file.
- [ ] Documentation in `.rst` files is written using [semantic newlines](https://rhodesmill.org/brandon/2012/one-sentence-per-line/).
- [ ] Changes (and possible deprecations) have news fragments in [`changelog.d`](https://github.com/python-attrs/attrs/blob/main/changelog.d).
- [x] Consider granting [push permissions to the PR branch](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), so maintainers can fix minor issues themselves without pestering you.

### CODE REVIEW

**Comment by Tinche** [comment]
I don't see a direct use-case for this (since attrs classes should be statically known now) but I don't see the harm either.

**Comment by layday** [comment]
The class might not be statically known to be an attrs class (#987, #996) or it can appear in a union with other types.

**Comment by hynek** [comment]
So what exactly does this do?

**Comment by Tinche** [comment]
It essentially makes `has` work at compile time too.

Let's say your function takes a class (not an instance, but an actual class). You will type your argument as `type` (or `type[C]`). You can check if it's an attrs class with `has`, and if that's true you can pass it to a different function that takes `type[AttrsInstance]`. But Mypy will still yell at you, since it doesn't know that `has` means it's an attrs class.

If we annotate `has` with `TypeGuard`, Mypy will know.

Code example:

```
def needs_attrs_class(cls: type[AttrsInstance]) -> None:
    pass

def my_function(cls: type) -> None:
    if has(cls):
        needs_attrs_class(cls)  # Mypy's gonna be unhappy, unless it knows `has` narrows the type of is argument
```

**Comment by layday** [comment]
It tells the type checker that if the condition is true then `cls` is of type `type[AttrsInstance]`. Simplified example:

```py
def is_dict_no_typeguard(value: object) -> bool:
    return isinstance(value, dict)

def is_dict_typeguard(value: object) -> TypeGuard[dict]:
    return isinstance(value, dict)

might_be_a_dict = ...

if is_dict_no_typeguard(might_be_a_dict):
    "type of might_be_a_dict is object"

if is_dict_typeguard(might_be_a_dict):
    "type of might_be_a_dict is dict"


```

**Comment by hynek** [comment]
Sounds cool! Any reason against it? If not: make it happen Tin. ;)

File: src/attr/__init__.pyi [hunk 1]
```diff
@@ -470,7 +470,15 @@ def astuple(
     tuple_factory: Type[Sequence[Any]] = ...,
     retain_collection_types: bool = ...,
 ) -> Tuple[Any, ...]: ...
-def has(cls: type) -> bool: ...
+
+if sys.version_info >= (3, 10):
```
Comment (hynek) [inline src/attr/__init__.pyi:474]: Would it be feasible to add a conditional import from typing-extensions?

Comment (layday) [reply src/attr/__init__.pyi:474]: You can only use one of a small number of constants like `TYPE_CHECKING` and `sys.version_info` to define types conditionally, try/except won't work, so I don't think so.

Comment (Tinche) [reply src/attr/__init__.pyi:474]: In my very limited testing this seems to work, both in Mypy and Pyright:

Screenshot 2022-08-15 at 13 37 05

Comment (layday) [reply src/attr/__init__.pyi:474]: Right, but typing_extensions is not a dependency of attrs, so that'll error if typing_extensions is not installed.

Comment (layday) [reply src/attr/__init__.pyi:474]: Okay, it seems typeshed ships typing_extensions is part of the stdlib, so I guess we can actually do that.

Comment (layday) [reply src/attr/__init__.pyi]: Done in https://github.com/python-attrs/attrs/pull/997/commits/fd2c819ab04bbfe4785ee6abbc9c42a8f566dc22.

**Comment by hynek** [comment]
Would you mind coming up with a way to verify this works in https://github.com/python-attrs/attrs/blob/main/tests/typing_example.py ?

**Comment by layday** [comment]
There's a test in `test_mypy.yml`.  Should I move it over to `typing_example.py`?

**Comment by hynek** [comment]
No, both would be great, since they work somewhat differently. Maybe `if has(c): print(c.__attrs_attrs__)` or something like that? That should pass, right?

**Review by Tinche — APPROVED** [review]

**Comment by hynek** [comment]
Just so I understand this correctly: the import from typing_extensions only works, because we have the stubs in pyi files that are only touched by mypy which in turn brings in typing_extensions? IOW: if we'd use inline type hints, this would fall apart?

**Comment by layday** [comment]
That's right. The typeshed has [`typing_extensions`](https://github.com/python/typeshed/blob/master/stdlib/typing_extensions.pyi) in the stdlib.  I assume they did this so that they can use new typing constructs in the typeshed without tripping MyPy.  At runtime, you'd have to do something like:

```py
import typing

if typing.TYPE_CHECKING:
    from typing_extensions import TypeGuard

# Use TypeGuard either in quotes or with `from __future__ import annotations`
```

**Comment by hynek** [comment]
Thanks!

**Pull request merged**